eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1408 @@
+/* Copyright 2018 Endless Mobile, Inc. */
+
+#include <string.h>
+#include <stdlib.h>
+
+#include <endless/endless.h>
+#include <json-glib/json-glib.h>
+#include <mustache.h>
+#include <libintl.h>
+
+#include "eknr-boxed-set-info.h"
+#include "eknr-errors.h"
+#include "eknr-renderer.h"
+
+#define _(String) gettext (String)
+#define gettext_noop(String) String
+#define N_(String) gettext_noop (String)
+
+/**
+ * SECTION:renderer
+ * @title: Renderer
+ * @short_description: A post-processing renderer for HTML article content
+ *
+ * The renderer is responsible for adding some final postprocessing to
+ * articles on the client side before it is displayed to the user. What
+ * postprocessing happens depends on the article's source.
+ */
+struct _EknrRenderer
+{
+  GObject parent_instance;
+};
+
+typedef struct _EknrRendererPrivate
+{
+  GHashTable *cache;
+} EknrRendererPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (EknrRenderer,
+                            eknr_renderer,
+                            G_TYPE_OBJECT)
+
+enum {
+  PROP_0,
+  NPROPS
+};
+
+static GParamSpec *eknr_renderer_props [NPROPS] = { NULL, };
+
+static mustache_str_ctx *
+_renderer_mustache_str_ctx_new (gchar *string)
+{
+  mustache_str_ctx *ctx = g_new0 (mustache_str_ctx, 1);
+  ctx->string = string;
+  ctx->offset = 0;
+
+  return ctx;
+}
+
+static void
+_renderer_mustache_str_ctx_free (mustache_str_ctx *ctx)
+{
+  g_clear_pointer (&ctx->string, g_free);
+  g_free (ctx);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (mustache_str_ctx, _renderer_mustache_str_ctx_free);
+
+typedef struct _RendererMustacheData {
+  GVariantDict  *variables;
+  GError       **error; /* non-owned */
+
+  mustache_str_ctx input;
+  mustache_str_ctx output;
+
+  const gchar *section_variable;
+} RendererMustacheData;
+
+static RendererMustacheData *
+renderer_mustache_data_new (GVariantDict  *variables,
+                            GError       **error,
+                            const gchar   *input)
+{
+  RendererMustacheData *data = g_new0 (RendererMustacheData, 1);
+  data->variables = variables;
+  data->error = error;
+  data->input.string = (gchar *) input;
```

Comment:
I'm not sure if `RendererMustacheData` should take a copy of the input string here - I suppose it could, though the way that we use this at the moment means does not cause any lifetime issues.